pkg/controller/agent/vlanconfig/controller.go

```diff
@@ -0,0 +1,402 @@
+		return nil, nil
+	}
+	klog.Infof("vlan config %s has been changed, spec: %+v", vc.Name, vc.Spec)
+
+	ok, v, err := h.MatchNode(vc)
+	if err != nil {
+		return nil, err
+	}
+	// Not match
+	if !ok {
+		if err := h.removeVLAN(vc); err != nil {
+			return nil, err
+		}
+		return vc, nil
+	}
+	// Another vlanConfig has take effect on this node
+	if v != "" && v != vc.Name {
+		klog.Infof("vlanConfig %s matches the node %s but goes after %s", vc.Name, h.nodeName, v)
+		return vc, nil
+	}
+	// set up VLAN
+	if err := h.setupVLAN(vc); err != nil {
+		return nil, err
+	}
+
+	return vc, nil
+}
+
+func (h Handler) OnRemove(_ string, vc *networkv1.VlanConfig) (*networkv1.VlanConfig, error) {
+	klog.Infof("vlan config %s has been removed", vc.Name)
+
+	if err := h.removeVLAN(vc); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (h Handler) MatchNode(vc *networkv1.VlanConfig) (bool, string, error) {
+	var matchedNodes []string
+	if err := json.Unmarshal([]byte(vc.Annotations[utils.KeyMatchedNodes]), &matchedNodes); err != nil {
+		return false, "", nil
+	}
+
+	for _, n := range matchedNodes {
+		if n == h.nodeName {
+			node, err := h.nodeCache.Get(n)
+			if err != nil {
+				return false, "", err
+			}
+			key := network.GroupName + "/" + vc.Spec.ClusterNetwork
+			return true, node.Labels[key], nil
+		}
+	}
+
+	return false, "", nil
+}
+
+func (h Handler) setupVLAN(vc *networkv1.VlanConfig) error {
+	var v *vlan.Vlan
+	var setupErr error
+	var vids []uint16
+	var uplink *iface.Link
+	// get VIDs
+	vids, setupErr = h.getNadVidList(iface.GenerateName(vc.Spec.ClusterNetwork, iface.BridgeSuffix))
+	if setupErr != nil {
+		goto updateStatus
+	}
+	// construct uplink
+	uplink, setupErr = setUplink(vc)
+	if setupErr != nil {
+		goto updateStatus
+	}
+	// set up VLAN
+	v = vlan.NewVlan(vc.Spec.ClusterNetwork, vids)
+	setupErr = v.Setup(uplink)
+
+updateStatus:
+	// Update status and still return setup error if not nil
+	if err := h.updateStatus(vc, v, setupErr); err != nil {
+		return fmt.Errorf("update status into vlanstatus %s", h.statusName(vc.Name))
+	}
+	if setupErr != nil {
+		return fmt.Errorf("set up VLAN failed, vlanconfig: %s, node: %s, error: %w", vc.Name, h.nodeName, setupErr)
+	}
+	// update node labels for pod scheduling
+	if err := h.addNodeLabel(vc); err != nil {
+		return fmt.Errorf("add node label to node %s for vlanconfig %s failed, error: %w", h.nodeName, vc.Name, err)
+	}
+
+	return nil
+}
+
+func (h Handler) removeVLAN(vc *networkv1.VlanConfig) error {
+	var v *vlan.Vlan
+	var teardownErr error
+
+	v, teardownErr = vlan.GetVlan(vc.Spec.ClusterNetwork)
+	// We take it granted that `LinkNotFound` means the VLAN has been torn down.
+	if teardownErr != nil {
+		if errors.As(teardownErr, &netlink.LinkNotFoundError{}) {
+			teardownErr = nil
+		}
+		goto updateStatus
+	}
+	if teardownErr = v.Teardown(); teardownErr != nil {
+		goto updateStatus
+	}
+
+updateStatus:
+	if err := h.removeNodeLabel(vc); err != nil {
```

Comment:
similar,
`teardownErr` is the first root cause, it should be formatted into return error

meanwhile, `log` those import information ?